File: 1/150/EvaluateReversePolishNotation.cpp
/*
	https://leetcode.com/problems/evaluate-reverse-polish-notation/description/?envType=daily-question&envId=2024-01-30
*/

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        int res=0;
        stack<int>stk;
        unordered_map<string, string>mp;
        mp["+"]="+", mp["-"]="-", mp["*"]="*", mp["/"]="/";
        for(auto &vals: tokens){
            if(mp.find(vals)==mp.end()){
                stk.push(stoi(vals));
            } else{
                int curtp=stk.top();
                stk.pop();
                int nxt=stk.top();
                stk.pop();
                if(vals=="+"){
                    stk.push(curtp+nxt);
                } else if(vals=="-"){
                    stk.push(nxt-curtp);
                } else if(vals=="*"){
                    stk.push(nxt*curtp);
                } else{
                    stk.push(nxt/curtp);
                }
            }
        }
        return stk.top();
    }
};

File: 1/150/EvaluateReversePolishNotation.py
# https://leetcode.com/problems/evaluate-reverse-polish-notation/description/?envType=daily-question&envId=2024-01-30

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        res=0
        stk=[]
        mp={}
        mp["+"]="+"
        mp["-"]="-"
        mp["*"]="*"
        mp["/"]="/"
        sz=len(tokens)
        for i in range(0, sz, 1):
            if tokens[i] not in mp:
                stk.append(int(tokens[i]))
            else:
                curtkn=tokens[i]
                curtp=stk[len(stk)-1]
                del(stk[len(stk)-1])
                nxt=stk[len(stk)-1]
                del(stk[len(stk)-1])
                if curtkn=="+":
                    stk.append(curtp+nxt)
                elif curtkn=="-":
                    stk.append(nxt-curtp)
                elif curtkn=="*":
                    stk.append(nxt*curtp)
                else:
                    stk.append(int(nxt/curtp))
        return stk[len(stk)-1]
        